atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncClusterSessionImpl.java

```diff
@@ -0,0 +1,132 @@
+
+public final class AsyncClusterSessionImpl implements AsyncClusterSession {
+
+    private static final Logger log = LoggerFactory.getLogger(AsyncClusterSessionImpl.class);
+
+    private final String sessionName;
+    private final Executor executor;
+
+    private final ScheduledExecutorService healthCheckExecutor = PTExecutors.newSingleThreadScheduledExecutor(
+            new NamedThreadFactory(sessionName() + "-healthCheck", true /* daemon */));
+    private final PreparedStatement healthCheckStatement;
+    private final Cluster cluster;
+    private final Session session;
+
+    public static AsyncClusterSessionImpl create(String clusterName, Cluster cluster, Session session,
+            ThreadFactory threadFactory) {
+        return create(clusterName, cluster, session, Executors.newCachedThreadPool(threadFactory));
+    }
+
+    public static AsyncClusterSessionImpl create(String clusterName, Cluster cluster, Session session,
+            Executor executor) {
+        PreparedStatement healthCheckStatement = session.prepare("SELECT dateof(now()) FROM system.local ;");
+        return new AsyncClusterSessionImpl(clusterName, cluster, session, executor, healthCheckStatement).start();
+    }
+
+    private AsyncClusterSessionImpl(String sessionName, Cluster cluster, Session session, Executor executor,
+            PreparedStatement healthCheckStatement) {
+        this.sessionName = sessionName;
+        this.cluster = cluster;
+        this.session = session;
+        this.executor = executor;
+        this.healthCheckStatement = healthCheckStatement;
+    }
+
+    @Override
+    public Map<MetricName, Metric> usedCqlClusterMetrics() {
+        return KeyedStream.stream(
+                cluster.getMetrics()
+                        .getRegistry()
+                        .getMetrics())
+                .mapKeys(name -> MetricName.builder().safeName(name).build())
+                .collectToMap();
+    }
+
+    @Override
+    public String sessionName() {
+        return sessionName;
+    }
+
+    public void close() {
+        healthCheckExecutor.shutdownNow();
+        log.info("Shutting down health checker for cluster session {}", SafeArg.of("clusterSession", sessionName));
+        boolean shutdown = false;
+        try {
+            shutdown = healthCheckExecutor.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.error("Interrupted while shutting down health checker. Should not happen");
+            Thread.currentThread().interrupt();
+        }
+
+        if (!shutdown) {
+            log.error("Failed to shutdown health checker in a timely manner for {}",
+                    SafeArg.of("clusterSession", sessionName));
+        } else {
+            log.info("Shut down health checker for cluster session {}", SafeArg.of("clusterSession", sessionName));
+        }
+    }
+
+    private AsyncClusterSessionImpl start() {
+        healthCheckExecutor.scheduleAtFixedRate(() -> {
```

Comment:
note: scheduleWithFixedRate is generally considered dangerous if you have multiple tasks on the same threadpool. But here it's fine.